src/syntax/platform.rkt

```diff
@@ -273,46 +248,58 @@
-                             info
-                             [convs (append (syntax->list #'(cs ...)) (platform-info-convs info))]))]
-         [(#:conversions bad _ ...) (oops! "expected a conversion list" #'bad)]
-         [(#:conversions) (oops! "expected conversion list after keyword `#:conversions`" stx)]
-         [(impl-sig rest ...)
-          (loop
-           #'(rest ...)
-           (struct-copy platform-info info [impls (cons #'impl-sig (platform-info-impls info))]))]
-         [_ (oops! "bad syntax")]))]))
+    [(_ id cs ...)
+     (let ([if-cost #f] [default-cost #f] [optional? #f])
+       (let loop ([cs #'(cs ...)] [impls '()] [costs '()] [reprs '()] [repr-costs '()])
+         (syntax-case cs ()
+           [()
+            (let ([platform-id #'id])
+              (with-syntax ([platform-id platform-id]
+                            [(impls ...) (reverse impls)]
+                            [(costs ...) (reverse costs)]
+                            [(reprs ...) reprs]
+                            [(repr-costs ...) repr-costs]
+                            [if-cost if-cost]
+                            [default-cost default-cost]
+                            [optional? optional?])
+                #'(define platform-id
+                    (make-platform `([impls ,costs] ...)
+                                   `([reprs ,repr-costs] ...)
+                                   #:optional? optional?
+                                   #:if-cost if-cost
+                                   #:default-cost default-cost))))]
+           [(#:if-cost cost rest ...)
+            (cond
+              [if-cost (oops! "multiple #:if-cost clauses" stx)]
+              [else
+               (set! if-cost (platform/parse-if-cost #'cost))
+               (loop #'(rest ...) impls costs reprs repr-costs)])]
+           [(#:if-cost) (oops! "expected value after keyword `#:if-cost`" stx)]
+           [(#:default-cost cost rest ...)
+            (cond
+              [if-cost (oops! "multiple #:default-cost clauses" stx)]
+              [else
+               (set! default-cost #'cost)
+               (loop #'(rest ...) impls costs reprs repr-costs)])]
+           [(#:default-cost) (oops! "expected value after keyword `#:default-cost`" stx)]
+           [(#:optional rest ...)
+            (cond
+              [optional? (oops! "multiple #:optional clauses" stx)]
+              [else
+               (set! optional? #t)
+               (loop #'(rest ...) impls costs reprs repr-costs)])]
+           [(#:literal [repr cost] rest ...)
+            (loop #'(rest ...) impls costs (cons #'repr reprs) (cons #'cost repr-costs))]
+           [(#:literals) (oops! "expected literals list after keyword `#:literals`" stx)]
+           [([impl cost] rest ...)
+            (loop #'(rest ...) (cons #'impl impls) (cons #'cost costs) reprs repr-costs)]
+           [(impl rest ...) (loop #'(rest ...) (cons #'impl impls) (cons #f costs) reprs repr-costs)]
```

Comment:
We should probably check that `impl` is not a keyword since this seems to be a default case. You can imagine having a typo and getting an unhelpful error that `#:keyword` is not an operator or something. I'm not entirely sure how to implement the check tbh.